File: app/src/main/java/medlife/com/br/adapter/CartAdapter.java
package medlife.com.br.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;
import java.util.List;
import medlife.com.br.R;
import medlife.com.br.model.CartItem;
import medlife.com.br.helper.CartManager;
import medlife.com.br.model.Farmacia;
import java.util.LinkedHashMap;
import java.util.ArrayList;
import java.util.Map;

public class CartAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {
    private static final int VIEW_TYPE_HEADER = 0;
    private static final int VIEW_TYPE_ITEM = 1;
    private Context context;
    private List<Object> displayList; // Can be Farmacia (header) or CartItem
    private CartListener listener;

    public interface CartListener {
        void onCartUpdated();
    }

    public CartAdapter(Context context, List<CartItem> cartItems, CartListener listener) {
        this.context = context;
        this.listener = listener;
        this.displayList = buildDisplayList(cartItems);
    }

    private List<Object> buildDisplayList(List<CartItem> cartItems) {
        Map<Farmacia, List<CartItem>> grouped = new LinkedHashMap<>();
        for (CartItem item : cartItems) {
            Farmacia farmacia = item.getProduct().getFarmacia();
            if (!grouped.containsKey(farmacia)) {
                grouped.put(farmacia, new ArrayList<>());
            }
            grouped.get(farmacia).add(item);
        }
        List<Object> result = new ArrayList<>();
        for (Map.Entry<Farmacia, List<CartItem>> entry : grouped.entrySet()) {
            result.add(entry.getKey());
            result.addAll(entry.getValue());
        }
        return result;
    }

    @Override
    public int getItemViewType(int position) {
        return (displayList.get(position) instanceof Farmacia) ? VIEW_TYPE_HEADER : VIEW_TYPE_ITEM;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        if (viewType == VIEW_TYPE_HEADER) {
            View view = LayoutInflater.from(context).inflate(R.layout.item_cart_pharmacy_header, parent, false);
            return new PharmacyHeaderViewHolder(view);
        } else {
            View view = LayoutInflater.from(context).inflate(R.layout.item_cart, parent, false);
            return new CartViewHolder(view);
        }
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        if (getItemViewType(position) == VIEW_TYPE_HEADER) {
            Farmacia farmacia = (Farmacia) displayList.get(position);
            PharmacyHeaderViewHolder headerHolder = (PharmacyHeaderViewHolder) holder;
            headerHolder.textPharmacyName.setText(farmacia.getName());
            headerHolder.textPharmacyLocation.setText(farmacia.getLocation());
            // Set logo based on pharmacy name (mock logic)
            if (farmacia.getName().toLowerCase().contains("são paulo")) {
                headerHolder.imagePharmacyLogo.setImageResource(R.drawable.mock_drogariasaopaulo);
            } else if (farmacia.getName().toLowerCase().contains("drogasil")) {
                headerHolder.imagePharmacyLogo.setImageResource(R.drawable.mock_logo_drogasil_2048);
            } else {
                headerHolder.imagePharmacyLogo.setImageResource(R.drawable.logo);
            }
        } else {
            CartItem cartItem = (CartItem) displayList.get(position);
            CartViewHolder itemHolder = (CartViewHolder) holder;
            itemHolder.productImage.setImageResource(cartItem.getProduct().getImage());
            itemHolder.productName.setText(cartItem.getProduct().getName());
            itemHolder.productPrice.setText(cartItem.getProduct().getPrice());
            itemHolder.quantityText.setText(String.valueOf(cartItem.getQuantity()));

            itemHolder.plusButton.setOnClickListener(v -> {
                CartManager.getInstance().updateQuantity(cartItem, cartItem.getQuantity() + 1);
                notifyDataSetChanged();
                listener.onCartUpdated();
            });

            itemHolder.minusButton.setOnClickListener(v -> {
                CartManager.getInstance().updateQuantity(cartItem, cartItem.getQuantity() - 1);
                notifyDataSetChanged();
                listener.onCartUpdated();
            });

            itemHolder.deleteButton.setOnClickListener(v -> {
                CartManager.getInstance().removeProduct(cartItem);
                notifyDataSetChanged();
                listener.onCartUpdated();
            });
        }
    }

    @Override
    public int getItemCount() {
        return displayList.size();
    }

    public static class CartViewHolder extends RecyclerView.ViewHolder {
        ImageView productImage;
        TextView productName;
        TextView productPrice;
        TextView quantityText;
        ImageView plusButton;
        ImageView minusButton;
        ImageView deleteButton;

        public CartViewHolder(@NonNull View itemView) {
            super(itemView);
            productImage = itemView.findViewById(R.id.product_image);
            productName = itemView.findViewById(R.id.product_name);
            productPrice = itemView.findViewById(R.id.product_price);
            quantityText = itemView.findViewById(R.id.quantity_text);
            plusButton = itemView.findViewById(R.id.plus_button);
            minusButton = itemView.findViewById(R.id.minus_button);
            deleteButton = itemView.findViewById(R.id.delete_button);
        }
    }

    public static class PharmacyHeaderViewHolder extends RecyclerView.ViewHolder {
        TextView textPharmacyName;
        TextView textPharmacyLocation;
        ImageView imagePharmacyLogo;
        public PharmacyHeaderViewHolder(@NonNull View itemView) {
            super(itemView);
            textPharmacyName = itemView.findViewById(R.id.textPharmacyName);
            textPharmacyLocation = itemView.findViewById(R.id.textPharmacyLocation);
            imagePharmacyLogo = itemView.findViewById(R.id.imagePharmacyLogo);
        }
    }

    public void updateCartItems(List<CartItem> cartItems) {
        this.displayList = buildDisplayList(cartItems);
        notifyDataSetChanged();
    }
} 

File: app/src/main/java/medlife/com/br/adapter/CartAdapter.kt
package medlife.com.br.adapter

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import medlife.com.br.R
import medlife.com.br.helper.CartManager
import medlife.com.br.model.CartItem
import medlife.com.br.model.Farmacia

class CartAdapter(
    private val context: Context,
    cartItems: List<CartItem>,
    private val listener: CartListener
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    private var displayList: MutableList<Any> = buildDisplayList(cartItems)

    interface CartListener {
        fun onCartUpdated()
    }

    private fun buildDisplayList(cartItems: List<CartItem>): MutableList<Any> {
        val grouped = LinkedHashMap<Farmacia, MutableList<CartItem>>()
        for (item in cartItems) {
            val product = item.product
            if (product != null) {
                val farmacia = product.farmacia
                if (farmacia != null) {
                    if (!grouped.containsKey(farmacia)) {
                        grouped[farmacia] = mutableListOf()
                    }
                    grouped[farmacia]?.add(item)
                }
            }
        }
        val result = mutableListOf<Any>()
        for ((key, value) in grouped) {
            result.add(key)
            result.addAll(value)
        }
        return result
    }

    override fun getItemViewType(position: Int): Int {
        return if (displayList[position] is Farmacia) VIEW_TYPE_HEADER else VIEW_TYPE_ITEM
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return if (viewType == VIEW_TYPE_HEADER) {
            val view = LayoutInflater.from(context).inflate(R.layout.item_cart_pharmacy_header, parent, false)
            PharmacyHeaderViewHolder(view)
        } else {
            val view = LayoutInflater.from(context).inflate(R.layout.item_cart, parent, false)
            CartViewHolder(view)
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (getItemViewType(position) == VIEW_TYPE_HEADER) {
            val farmacia = displayList[position] as Farmacia
            val headerHolder = holder as PharmacyHeaderViewHolder
            headerHolder.textPharmacyName.text = farmacia.name
            headerHolder.textPharmacyLocation.text = farmacia.location
            
            val nameLower = farmacia.name?.lowercase() ?: ""
            if (nameLower.contains("são paulo")) {
                headerHolder.imagePharmacyLogo.setImageResource(R.drawable.mock_drogariasaopaulo)
            } else if (nameLower.contains("drogasil")) {
                headerHolder.imagePharmacyLogo.setImageResource(R.drawable.mock_logo_drogasil_2048)
            } else {
                headerHolder.imagePharmacyLogo.setImageResource(R.drawable.logo)
            }
        } else {
            val cartItem = displayList[position] as CartItem
            val itemHolder = holder as CartViewHolder
            val product = cartItem.product
            
            if (product != null) {
                itemHolder.productImage.setImageResource(product.image)
                itemHolder.productName.text = product.name
                itemHolder.productPrice.text = product.price
            }
            itemHolder.quantityText.text = cartItem.quantity.toString()

            itemHolder.plusButton.setOnClickListener {
                CartManager.updateQuantity(cartItem, cartItem.quantity + 1)
                notifyDataSetChanged()
                listener.onCartUpdated()
            }

            itemHolder.minusButton.setOnClickListener {
                CartManager.updateQuantity(cartItem, cartItem.quantity - 1)
                notifyDataSetChanged()
                listener.onCartUpdated()
            }

            itemHolder.deleteButton.setOnClickListener {
                CartManager.removeProduct(cartItem)
                notifyDataSetChanged()
                listener.onCartUpdated()
            }
        }
    }

    override fun getItemCount(): Int {
        return displayList.size
    }

    fun updateCartItems(cartItems: List<CartItem>) {
        this.displayList = buildDisplayList(cartItems)
        notifyDataSetChanged()
    }

    class CartViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val productImage: ImageView = itemView.findViewById(R.id.product_image)
        val productName: TextView = itemView.findViewById(R.id.product_name)
        val productPrice: TextView = itemView.findViewById(R.id.product_price)
        val quantityText: TextView = itemView.findViewById(R.id.quantity_text)
        val plusButton: ImageView = itemView.findViewById(R.id.plus_button)
        val minusButton: ImageView = itemView.findViewById(R.id.minus_button)
        val deleteButton: ImageView = itemView.findViewById(R.id.delete_button)
    }

    class PharmacyHeaderViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val textPharmacyName: TextView = itemView.findViewById(R.id.textPharmacyName)
        val textPharmacyLocation: TextView = itemView.findViewById(R.id.textPharmacyLocation)
        val imagePharmacyLogo: ImageView = itemView.findViewById(R.id.imagePharmacyLogo)
    }

    companion object {
        private const val VIEW_TYPE_HEADER = 0
        private const val VIEW_TYPE_ITEM = 1
    }
}
